crates/autopilot/src/on_settlement_event_updater.rs

```diff
@@ -202,6 +209,65 @@ impl OnSettlementEventUpdater {
 
         Ok(true)
     }
+
+    /// With solver driver colocation solvers are supposed to append the
+    /// `auction_id` to the settlement calldata. This function tries to
+    /// recover that `auction_id`. This function only returns an error if
+    /// retrying the operation makes sense. If all went well and there
+    /// simply is no sensible `auction_id` `Ok(None)` will be returned.
+    async fn recover_auction_id_from_calldata(&self, tx: &Transaction) -> Result<Option<i64>> {
+        let metadata = match DecodedSettlement::new(&tx.input.0) {
+            Ok(settlement) => settlement.metadata,
+            Err(err) => {
+                tracing::warn!(
+                    ?tx,
+                    ?err,
+                    "could not decode settlement tx, unclear which auction it belongs to"
+                );
+                return Ok(None);
+            }
+        };
+        let auction_id = match metadata {
+            Some(bytes) => i64::from_be_bytes(bytes.0),
+            None => {
+                tracing::warn!(?tx, "could not recover the auction_id from the calldata");
+                return Ok(None);
+            }
+        };
+        let participants = database::auction_participants::fetch(
+            self.db
+                .0
+                .acquire()
+                .await
+                .context("acquire DB connection")?
+                .deref_mut(),
+            auction_id,
+        )
+        .await
+        .context("fetch auction participants")?;
+
+        if participants.is_empty() {
+            tracing::debug!(
+                auction_id,
+                ?tx,
+                "settlement likely originated from other environment"
+            );
+            Ok(None)
+        } else if participants
+            .iter()
+            .any(|p| Some(H160(p.participant.0)) == tx.from)
+        {
+            tracing::debug!(auction_id, "extracted auction_id from settlement metadata");
+            Ok(Some(auction_id))
+        } else {
+            tracing::warn!(
+                auction_id,
+                ?tx,
+                "settlement was submitted by solver that didn't participate in the competition"
+            );
```

Comment:
IMO this is a pretty serious issue so I would liked to have made this an `error` log. But because we have to environments (staging, prod) with 2 separate `auction_id` counters we may see `auction_id` collisions in the future which makes this check not bullet proof and might result in noisy false alarms.
Imagine 6 months from now prod reaches the same the current `auction_id` of staging. Then we settle an auction on prod and staging sees it. The staging DB might already have settled the same auction and would notice that the prod winner didn't even compete in the staging auction.

Reply:
I would make this an `error` TBH. We can instead encode additional metadata other than the auction ID (like a 4-byte "environment" selector as `b"prod"` or `b"barn"`).

We can even pass a `salt` in the `auction` or in the `/settle` API call to the driver that must be appended.

Reply:
With the `salt` we might as well go back to where we started. 😅  (autopilot generating a secret key the solver is supposed to append)
Although this would require another table which associates the `auction_id` to the `salt`.

Reply:
I also don't think they are likely to collide for any period of time.

Reply:
I think last time this came up I checked and estimated that the prod `auction_id` would catch up in staging in 6 months. Granted, this still doesn't guarantee that collisions will happen (not all `auction_id`s will lead to settlements) but the chance is there. Especially if we would really push for 1 block per auction.

Reply:
> estimated that the prod auction_id would catch up in staging in 6 months

At which point, `prod` would surpass staging right? So its unlikely to stay "noisy" for any period of time - it might collide for a settlement or two at most.

Reply:
Agree we should prevent this with code though - at which point, I do think we need some salt and, as you said:

> With the `salt` we might as well go back to where we started 😅

Reply:
Sorry, I meant it would catch up to the current staging ID. Meaning that in 6 months prod auction_ids will always have the potential to collide with a 6 months old staging ids.

Reply:
> And even if one environment actually overtakes the other one it would only move the potential for collisions into the other environment.

Right, but my point is, if we assume an average `auctions/secs`, that the collision possibility will only happen when the two "auction IDs / time lines" collide.

So, once prod overtakes staging, I imagine auction IDs will continue to grow at a quicker rate in production than in staging and make it so they don't collide anymore.

I agree that this is not ideal and we should solve it though.

Reply:
Collisions (at least in the context of this particular thread) are cases when e.g. staging indexes a settlement from prod but already has data for that auction_id (because it generated and settled that auction_id first). Now when staging looks at the participants for this auction it will not find the `tx_from` because it came from a prod solver which did not partake in the staging environment.
At this point it's unclear whether some rogue staging solver settled this auction out of order or if a law abiding solver from the other environment settled this auction fair and square.
This will never change. Not with both auction ids being close to each other nor with 1 environment being way ahead of the other.
The only thing we could do to decide what happened (rogue solver vs. settlement from other env) is to check if our env already indexed the auction data for this id. If we already know about the auction we could just assume the solution came from the other env.

Reply:
> check if our env already indexed the settlement_observation

Actually we can't even do that at the moment because the only thing associating an `auction_id` with a transaction is the `auction_transactions` table which we will no longer populate because we are missing the `nonce`. 😩

Reply:
[6590139](https://github.com/cowprotocol/services/pull/1675/commits/659013951d89d750b2be660a5f5f39ed9ce72b1b) and more importantly [7ba9e4b](https://github.com/cowprotocol/services/pull/1675/commits/7ba9e4bb0e5ec28c1319ff19f70b2fccb465ec15) try to address this issue as accurately as possible.
Could you give it another careful look to see if we are good @nlordell?
Another big issue I noticed is that theoretically a malicious solver could submit a solution with an auction_id for which we already recorded all the data. This could have resulted in us overwriting existing data.
Now we only insert into the `settlement_observations` table if the solution was provided by the winning solver for the given address and we don't already have data for that. To ensure correctness I had to do the entire update in a single big DB transaction.